samcli/lib/utils/version_checker.py

```diff
@@ -0,0 +1,128 @@
+DELTA_DAYS = 7
+
+
+def check_newer_version(func):
+    """
+    This function returns a wrapped function definition, which checks if there are newer version of SAM CLI available
+
+    Parameters
+    ----------
+    func: function reference
+        Actual function (command) which will be executed
+
+    Returns
+    -------
+    function reference:
+        A wrapped function reference which executes original function and checks newer version of SAM CLI
+    """
+
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        # execute actual command first
+        actual_result = func(*args, **kwargs)
+        # check and inform newer version if it is available
+        inform_newer_version()
+
+        return actual_result
+
+    return wrapped
+
+
+def inform_newer_version(force_check=False):
+    """
+    Compares installed SAM CLI version with the up to date version from PyPi,
+    and print information if up to date version is different then what is installed now
+
+    It will store last version check time into GlobalConfig, so that it won't be running all the time
+    Currently, it will be checking weekly
+
+    Parameters
+    ----------
+    check_all_always: bool
+        When it is True, it will trigger checking new version of SAM CLI. Default value is False
+
+    """
+    # run everything else in try-except block
+    global_config = None
+    need_to_update_last_check_time = True
+    try:
+        global_config = GlobalConfig()
+        last_version_check = global_config.last_version_check
+
+        if force_check or is_last_check_older_than_delta(last_version_check):
+            fetch_and_compare_versions()
+        else:
+            need_to_update_last_check_time = False
+    except Exception as e:
+        LOG.debug("New version check failed", exc_info=e)
+    finally:
+        if need_to_update_last_check_time:
+            update_last_check_time(global_config)
+
+
+def fetch_and_compare_versions():
+    """
+    Compare current up to date version with the installed one, and inform if a newer version available
+    """
+    response = get(AWS_SAM_CLI_PYPI_ENDPOINT, timeout=PYPI_CALL_TIMEOUT_IN_SECONDS)
+    result = response.json()
+    latest_version = result.get("info", {}).get("version", None)
+    LOG.debug("Installed version %s, current version %s", installed_version, latest_version)
+    if latest_version and installed_version != latest_version:
+        click.secho(f"\nSAM CLI update available ({latest_version}); ({installed_version} installed)", fg="green")
+        click.echo(f"To download: {AWS_SAM_CLI_INSTALL_DOCS}")
+
+
+def update_last_check_time(global_config):
+    """
+    Update last_check_time in GlobalConfig
+    Parameters
+    ----------
+    global_config: GlobalConfig
+        GlobalConfig object that have been read
+    """
+    try:
+        if global_config:
+            global_config.last_version_check = datetime.utcnow().timestamp()
+    except Exception as e:
```

Comment:
When does this fail?

Reply:
It updates value and writes it to the config file which is kept under `~/.aws-sam/metadata.json`. For any reason, I don't want this to fail original command with a non-zero exit code.

Reply:
Oh okay I see, interesting. So what looks like a simple assignment actually does I/O.

Reply:
Yeah, the setter is overridden inside `GlobalConfig` and it is doing all this magic there